scipy/sparse/data.py

```diff
@@ -64,8 +64,16 @@ def __itruediv__(self, other):  # self /= other
         else:
             return NotImplemented
 
-    def astype(self, t):
-        return self._with_data(self._deduped_data().astype(t))
+    def astype(self, dtype, casting='unsafe', copy=True):
+        dtype = np.dtype(dtype)
+        if self.dtype != dtype:
+            return self._with_data(
+                self._deduped_data().astype(dtype, casting=casting, copy=copy),
+                copy=False)
```

Comment:
The issue with using `copy=False` here is that it means we might end  up sharing index arrays between `self` and the result, even with `astype(..., copy=True)`. For example, with the current proposed changes:
```python
a = csr_matrix(np.eye(3), dtype=int)
b = a.astype(float, copy=True)
a.indices.base is b.indices.base  # ==> True, bad news!
```
Thankfully, `_with_data(..., copy=True)` only copies `self.indices` and `self.indptr`, so we aren't adding an extra copy of the data array.